pkg/logicalrestore/logicalrestore.go

```diff
@@ -241,94 +199,214 @@ func (r *LogicalRestore) applyDelta(filePath string) error {
 		msg, err := decoder.Parse(dataBuf)
 		if err != nil {
 			return fmt.Errorf("could not parse message: %v", err)
-			break
 		}
 
 		switch v := msg.(type) {
 		case message.Relation:
-			//TODO: mind rename table case
-			if r.relInfo.Name != v.Name || r.relInfo.Namespace != v.Namespace {
-				log.Printf("message does not belong to current relation")
-			}
-			sql = v.SQL(r.relInfo)
-			r.relInfo.Columns = v.Columns
-			r.relInfo.ReplicaIdentity = v.ReplicaIdentity
-		case message.Insert:
-			sql = v.SQL(r.relInfo)
-		case message.Update:
-			sql = v.SQL(r.relInfo)
-		case message.Delete:
-			sql = v.SQL(r.relInfo)
+			r.relInfo = v
 		case message.Begin:
-			curTx = v.XID
-			sql = fmt.Sprintf("BEGIN -- %d", curTx)
+			if r.tx != nil {
+				panic("there is tx already open")
+			}
+
+			if v.FinalLSN < r.startLSN {
+				r.curLsn = dbutils.InvalidLsn
+				r.curTx = 0
+				break
```

Comment:
I think it makes sense to skip everything until the end of the current transaction (even before entering the switch).